paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/CheckCloneResultAndCopySnapshotOperator.java

```diff
@@ -0,0 +1,192 @@
+        //                        + cloneFileInfo.getFileSize()
+        //                        + ", target file path is : "
+        //                        + targetFilePath
+        //                        + ", and target file size is  : "
+        //                        + targetTableFileIO.getFileSize(targetFilePath));
+    }
+
+    @Override
+    public void finish() throws Exception {
+        super.finish();
+        copySnapshotOrTag();
+        commitSnapshotHintInTargetTable();
+    }
+
+    @Override
+    public void close() throws Exception {
+        super.close();
+        if (snapshotOrTagFiles != null) {
+            snapshotOrTagFiles.clear();
+        }
+    }
+
+    /**
+     * copy snapshot or tag in this latest operator of flink batch job.
+     *
+     * @throws IOException
+     */
+    private void copySnapshotOrTag() throws IOException {
+        for (CloneFileInfo cloneFileInfo : snapshotOrTagFiles) {
+            CopyFileUtils.copyFile(
+                    cloneFileInfo,
+                    sourceTableFileIO,
+                    targetTableFileIO,
+                    sourceTableRootPath,
+                    targetTableRootPath,
+                    copyInDifferentCluster);
+        }
+    }
+
+    /**
+     * commit Snapshot earliestHint and latestHint in targetTable in different cloneTypes.
+     *
+     * @throws IOException
+     */
+    private void commitSnapshotHintInTargetTable() throws IOException {
+        switch (cloneType) {
+            case LatestSnapshot:
+            case SpecificSnapshot:
+            case FromTimestamp:
+                checkState(
+                        1 == targetTableSnapshotManager.snapshotCount(),
+                        "snapshot count is not equal to 1 "
+                                + "when cloneType is LATEST_SNAPSHOT / SPECIFIC_SNAPSHOT / FROM_TIMESTAMP.");
+                long snapshotId = targetTableSnapshotManager.safelyGetAllSnapshots().get(0).id();
+                targetTableSnapshotManager.commitEarliestHint(snapshotId);
+                targetTableSnapshotManager.commitLatestHint(snapshotId);
+                break;
+
+            case Table:
+                long maxSnapshotId = Long.MIN_VALUE;
+                long minSnapshotId = Long.MAX_VALUE;
+                List<Snapshot> snapshots = targetTableSnapshotManager.safelyGetAllSnapshots();
+                for (Snapshot snapshot : snapshots) {
+                    maxSnapshotId = Math.max(maxSnapshotId, snapshot.id());
+                    minSnapshotId = Math.min(minSnapshotId, snapshot.id());
+                }
+                targetTableSnapshotManager.commitEarliestHint(minSnapshotId);
+                targetTableSnapshotManager.commitLatestHint(maxSnapshotId);
+                break;
+
+            case Tag:
+            default:
+                throw new UnsupportedOperationException("Unknown cloneType : " + cloneType);
```

Comment:
Does not support tag?

Reply:
My opinion is If user clone tag, then no snapshot file in snapshot dir of target table.
Because I see the code, when expire snapshot, if a snapshot tagged, then snapshot file alse be deleted.